llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -10195,12 +10193,38 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
       return VecCost;
     };
     if (SLPReVec && !E->isAltShuffle())
-      return GetCostDiff(GetScalarCost, [](InstructionCost) {
-        // shufflevector will be eliminated by instcombine because the
-        // shufflevector masks are used in order (guaranteed by
-        // getShufflevectorNumGroups). The vector cost is 0.
-        return TTI::TCC_Free;
-      });
+      return GetCostDiff(
+          GetScalarCost, [&](InstructionCost) -> InstructionCost {
+            // If a group uses mask in order, the shufflevector can be
+            // eliminated by instcombine. Then the cost is 0.
+            bool IsIdentity = true;
+            auto *SV = cast<ShuffleVectorInst>(VL.front());
+            unsigned SVNumElements =
+                cast<FixedVectorType>(SV->getOperand(0)->getType())
+                    ->getNumElements();
+            unsigned GroupSize = SVNumElements / SV->getShuffleMask().size();
+            for (size_t I = 0, E = VL.size(); I != E; I += GroupSize) {
+              ArrayRef<Value *> Group = VL.slice(I, GroupSize);
+              int NextIndex = 0;
+              if (!all_of(Group, [&](Value *V) {
+                    auto *SV = cast<ShuffleVectorInst>(V);
+                    int Index;
+                    SV->isExtractSubvectorMask(Index);
+                    if (NextIndex != Index)
+                      return false;
```

Comment:
Actually, I think better just to include the cost of the vector extract here, rather than consider it free. Better to request the cost from TTI. Also, check that `SV->isExtractSubvectorMask(Index);` returns true

Reply:
When do we need to calculate the vector extract here? Vectorization will merge multiple shufflevector into 1 shufflevector. Vector extract is not used.

Reply:
Fix in https://github.com/llvm/llvm-project/pull/106212/commits/25b901814eef9dd5716130a0ee31adc5cac124c6.